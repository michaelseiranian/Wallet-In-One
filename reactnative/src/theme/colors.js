// Colours used across different components of the app when 'light' theme is enabled in settings.
export const lightColors = {
    background: 'white',
    text: 'black',
    primary: 'purple',  // Main theme colour of the app
    secondary: 'blue',
    stock_account: 'white', // Used for the background of the main stocks screen
};

// Colours used across different components of the app when 'dark' theme is enabled in settings.
export const darkColors = {
    background: '#242526',
    text: 'white',
    primary: 'purple',  // Main theme colour of the app
    secondary: 'blue',
    stock_account: '#181818', // Used for the background of the main stocks screen
};