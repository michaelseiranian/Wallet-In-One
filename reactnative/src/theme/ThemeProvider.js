import React, {createContext, useEffect, useContext, useState, AsyncStorage} from 'react';
import {lightColors, darkColors} from './colors';
import {useColorScheme} from 'react-native';
export const ThemeContext = createContext({
    dark: false,
    colors: lightColors,
    setScheme: () => {},
    update: () => {}
});

import * as SecureStore from 'expo-secure-store';

export const ThemeProvider = props => {
    const defaultColorScheme = useColorScheme(); // colorScheme is either 'dark', 'light' or 'null'
    const [isDark, setIsDark] = useState(false);

    async function loadDarkModeSettings(value = 'default') {
        // colorScheme is either 'dark', 'light' or 'null'
        var darkSetting;
        if (value === "default"){
            darkSetting = await SecureStore.getItemAsync('darkModeSettings');
        }
        else{
            darkSetting = value;
        }

        console.log('settings, default: ', darkSetting, defaultColorScheme,value)

        if (darkSetting === 'true') {
            setIsDark(true)
        }
        else if (darkSetting === 'false') {
            setIsDark(false)
        }
        else if (defaultColorScheme == 'dark') {
            setIsDark(true)
        }
        else if (defaultColorScheme == 'light') {
            setIsDark(false)
        }
        else {
            setIsDark(false)
        }
    }

    // Load Dark Mode Setting
    useEffect(() => {    
        loadDarkModeSettings();
    }, [defaultColorScheme]);

    const defaultTheme = {
        dark: isDark,
        colors: isDark ? darkColors : lightColors,
        setScheme: scheme => setIsDark(scheme === 'dark'),
        update: loadDarkModeSettings,
    };
    return (
        <ThemeContext.Provider value={defaultTheme}>
            {props.children}
        </ThemeContext.Provider>
    );
};

// creating a custom hook for accessing all the values

export const useTheme = () => useContext(ThemeContext);