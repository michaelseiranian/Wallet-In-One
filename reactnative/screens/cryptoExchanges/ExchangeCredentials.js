import React, { useState } from 'react';
import { StyleSheet, Pressable, View, Text, TextInput, Button, Alert } from 'react-native';
import * as SecureStore from "expo-secure-store";
import { useTheme } from 'reactnative/src/theme/ThemeProvider';
import { api_url } from '../../authentication';


export default function ExchangeCredentials({ route, navigation }) {
  const [apiKey, setApiKey] = useState('');
  const [secretKey, setSecretKey] = useState('');
  const {dark, colors, setScheme} = useTheme();
  const {exchange} = route.params;

  // The following function is triggered on press of Submit button
  const handleSubmit = async () => {
    // If fields are empty, display appropriate error message
    if (!apiKey || !secretKey) {
      Alert.alert('Error', 'Please enter both API Key and Secret Key.');
      return;
    }
    // else, try to find the exchange account and Post into backend
    try {
      const response = await fetch(`${api_url}/crypto-exchanges/${exchange.toLowerCase()}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
        },
        body: JSON.stringify({ api_key: apiKey, secret_key: secretKey }),
      });
      const data = await response.json();
      const statusCode = response.status;
      // Success
      if (statusCode === 200) {
        navigation.navigate('Crypto Wallets & Exchanges');
      }
      // Error posting
      else {
        Alert.alert('Error', data["error"]);
      }
      // Any other error thrown in the try block
    } catch (error) {
      Alert.alert('Error', `An error occurred while retrieving ${exchange} account data.`);
    }
  };

  const styles = StyleSheet.create({
    titleContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      marginBottom: 20,
    },
    title: {
      fontSize: 25,
      color: colors.text,
    },
    input:{
      height: 40,
      width: '100%',
      borderWidth: 0.5,
      padding: 10,
      borderColor: 'gray',
      borderRadius: 5,
      marginTop: 5,
      marginBottom: 10,
      color: colors.text,
      backgroundColor: colors.background

    },
  });

  return (
    <View style={{ padding: 20, backgroundColor:colors.background, flex: 1 }}>
      <View style={{ flexDirection: "row", alignItems: "center", marginBottom: 20 }}>
        <Text style={styles.title}>{exchange} Credentials:</Text>
      </View>
      {/*API key title and field*/}
      <Text style={{ fontSize: 20, marginBottom: 10, color: colors.text }}>API Key:</Text>
      <TextInput
        autoCapitalize='none'
        autoCorrect={false}
        value={apiKey}
        onChangeText={setApiKey}
        style={styles.input}
        testID="apiKeyInput"
      />
      {/*Secret key title and field*/}
      <Text style={{ fontSize: 20, marginBottom: 10, color: colors.text }}>Secret Key:</Text>
      <TextInput
        autoCapitalize='none'
        autoCorrect={false}
        value={secretKey}
        onChangeText={setSecretKey}
        secureTextEntry
        style={styles.input}
        testID="secretKeyInput"
      />
      {/*Submit button*/}
      <Button
        title="Submit"
        onPress={handleSubmit}
        color= {colors.primary}
        backgroundColor='#FFFF00'
        buttonStyle={{ borderRadius: 20 }}
      />
    </View>
  );
}
