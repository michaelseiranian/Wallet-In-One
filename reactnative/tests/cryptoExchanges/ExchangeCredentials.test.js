import React from 'react';
import {Alert} from "react-native";
import {act, fireEvent, render} from '@testing-library/react-native';
import ExchangeCredentials from "../../screens/cryptoExchanges/ExchangeCredentials";
import * as SecureStore from "expo-secure-store";
import {useTheme} from 'reactnative/src/theme/ThemeProvider';
import {describe, it} from "@jest/globals";
import expect from "expect";


jest.mock("expo-secure-store");
jest.mock('reactnative/src/theme/ThemeProvider');

const mockNavigation = {
    navigate: jest.fn(),
};

const mockRoute = {
    params: {
        exchange: 'TestExchange',
    },
};

describe('ExchangeCredentials', () => {
    const props = {
        route: {
            params: {
                exchange: 'binance',
                id: 1
            },

        },
        navigation: {
            goBack: jest.fn(),
            navigate: jest.fn()
        }
    };

    it('renders correctly snapshot', () => {
        const { toJSON } = render(<ExchangeCredentials {...props} />);
        expect(toJSON()).toMatchSnapshot();
    });

    beforeEach(() => {
        SecureStore.getItemAsync.mockReset();
        useTheme.mockReturnValue({
            dark: false,
            colors: {
                text: '#000',
                background: '#fff',
                primary: '#0af',
            },
            setScheme: jest.fn(),
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { getByText, getByTestId } = render(
            <ExchangeCredentials route={mockRoute} navigation={mockNavigation} />,
        );

        expect(getByText('TestExchange Credentials:')).toBeTruthy();
        expect(getByText('API Key:')).toBeTruthy();
        expect(getByText('Secret Key:')).toBeTruthy();
    });

    it('shows an error if either API Key or Secret Key is missing', async () => {
        const { getByText, getByTestId } = render(
            <ExchangeCredentials route={mockRoute} navigation={mockNavigation} />,
        );

        jest.spyOn(Alert, 'alert');

        const apiKeyInput = getByTestId('apiKeyInput');
        const secretKeyInput = getByTestId('secretKeyInput');
        const submitButton = getByText('Submit');

        fireEvent.changeText(apiKeyInput, '');
        fireEvent.changeText(secretKeyInput, 'testSecretKey');
        fireEvent.press(submitButton);
        expect(Alert.alert).toHaveBeenCalledWith("Error",
            "Please enter both API Key and Secret Key.");

        fireEvent.changeText(apiKeyInput, 'testApiKey');
        fireEvent.changeText(secretKeyInput, '');
        fireEvent.press(submitButton);
        expect(Alert.alert).toHaveBeenCalledWith("Error",
            "Please enter both API Key and Secret Key.");
    });

    it('shows an error if wrong API/secret key pair is provided', async () => {
        const { getByText, getByTestId } = render(<ExchangeCredentials {...props} />);

        const apiKeyInput = getByTestId('apiKeyInput');
        const secretKeyInput = getByTestId('secretKeyInput');
        const submitButton = getByText('Submit');

        fireEvent.changeText(apiKeyInput, 'testAPIKey');
        fireEvent.changeText(secretKeyInput, 'testSecretKey');
        fireEvent.press(submitButton);
    });

    it('creates account successfully, with success message', async() => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                status: 200, json: () => Promise.resolve([{id: 1, name: 'Exchange 1'}]),
            })
        );

        const { getByText, getByTestId } = render(<ExchangeCredentials {...props }/>);

        const apiKeyInput = getByTestId('apiKeyInput');
        const secretKeyInput = getByTestId('secretKeyInput');
        const submitButton = getByText('Submit');

        fireEvent.changeText(apiKeyInput, 'testAPIKey');
        fireEvent.changeText(secretKeyInput, 'testSecretKey');
        fireEvent.press(submitButton);

    })

    it('throws error on status code 401', async() => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                status: 401, json: () => Promise.resolve([{id: 1, name: 'Exchange 1'}]),
            })
        );

        const { getByText, getByTestId } = render(<ExchangeCredentials {...props}/>);

        const apiKeyInput = getByTestId('apiKeyInput');
        const secretKeyInput = getByTestId('secretKeyInput');
        const submitButton = getByText('Submit');

        fireEvent.changeText(apiKeyInput, 'testAPIKey');
        fireEvent.changeText(secretKeyInput, 'testSecretKey');
        fireEvent.press(submitButton);
    })

    it('throws error on catch error', async() => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                status: 401, json: () => Promise.reject(),
            })
        );

        const { getByText, getByTestId } = render(<ExchangeCredentials {...props}/>);

        jest.spyOn(Alert, 'alert');
        const apiKeyInput = getByTestId('apiKeyInput');
        const secretKeyInput = getByTestId('secretKeyInput');
        const submitButton = getByText('Submit');

        fireEvent.changeText(apiKeyInput, 'testAPIKey');
        fireEvent.changeText(secretKeyInput, 'testSecretKey');

        await act(async () => {
            await fireEvent.press(submitButton);
        })
        fireEvent.press(submitButton);

        expect(Alert.alert).toHaveBeenCalledWith("Error", "An error occurred while retrieving binance account data.");
    })

});

