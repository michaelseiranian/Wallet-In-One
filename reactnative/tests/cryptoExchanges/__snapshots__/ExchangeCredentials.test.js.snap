// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeCredentials renders correctly snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 25,
        }
      }
    >
      binance
       Credentials:
    </Text>
  </View>
  <Text
    style={
      {
        "color": "#000",
        "fontSize": 20,
        "marginBottom": 10,
      }
    }
  >
    API Key:
  </Text>
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "gray",
        "borderRadius": 5,
        "borderWidth": 0.5,
        "color": "#000",
        "height": 40,
        "marginBottom": 10,
        "marginTop": 5,
        "padding": 10,
        "width": "100%",
      }
    }
    testID="apiKeyInput"
    value=""
  />
  <Text
    style={
      {
        "color": "#000",
        "fontSize": 20,
        "marginBottom": 10,
      }
    }
  >
    Secret Key:
  </Text>
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    onChangeText={[Function]}
    secureTextEntry={true}
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "gray",
        "borderRadius": 5,
        "borderWidth": 0.5,
        "color": "#000",
        "height": 40,
        "marginBottom": 10,
        "marginTop": 5,
        "padding": 10,
        "width": "100%",
      }
    }
    testID="secretKeyInput"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            {
              "color": "#0af",
            },
          ]
        }
      >
        Submit
      </Text>
    </View>
  </View>
</View>
`;
